rs/replicated_state/src/canister_state.rs

```diff
@@ -437,11 +439,11 @@ impl CanisterState {
             + NumBytes::from(self.system_state.certified_data.len() as u64)
     }
 
-    /// Sets the (transient) size in bytes of responses from this canister
-    /// routed into streams and not yet garbage collected.
-    pub(super) fn set_stream_responses_size_bytes(&mut self, size_bytes: usize) {
+    /// Sets the (transient) size in bytes of guaranteed responses from this
+    /// canister routed into streams and not yet garbage collected.
+    pub(super) fn set_stream_guaranteed_responses_size_bytes(&mut self, size_bytes: usize) {
```

Comment:
I'm not sure I understand. We do garbage collect messages in streams when we receive signals for them (meaning that they were "seen" by the destination subnet).

We are only interested in guaranteed responses, because they count against the guaranteed response memory limit.

Technically we could also aggregate the size of best-effort messages in streams, but that's not particularly useful, since we can no longer shed them. So if it e.g. happened that we had so many best-effort messages in streams that they alone exceeded the memory allowance for best-effort messages, there would be absolutely nothing we could do about that.

Going to the other extreme, considering that we could simply limit the total size of all streams (and even load-balance across streams), I'm wondering whether we shouldn't instead simply drop `stream_guaranteed_responses_size_bytes` altogether and only account the messages in a canister's queues against its guaranteed response message memory usage; with the streams a fixed "cost" per subnet. It would simplify quite a bit of logic and constraints around maintaining and updating these per-canister transient fields.

Reply:
> So if it e.g. happened that we had so many best-effort messages in streams that they alone exceeded the memory allowance for best-effort messages, there would be absolutely nothing we could do about that.

This was the scenario I had in mind and whether we'd want/need to do something about it. The argument that the streams are relatively small anyways is a good one. Thanks for the explanation.

> I'm wondering whether we shouldn't instead simply drop stream_guaranteed_responses_size_bytes

Sounds like a good option as well. I'd do this in a separate ticket though.

Reply:
> I'd do this in a separate ticket though.

Definitely.

Reply:
It does precisely track the message memory usage of each canister (so it can be charged for it). It does not always precisely track the total memory usage of the subnet, if a canister was deleted or migrated (although we could also do that, if we wanted that level of precision). A "transient field" is a field whose value is valid temporarily (i.e. for the time span between two rounds, in this context) and not persisted.

Regarding the reject responses we generate during induction, we could not charge a canister that has migrated for them (nor would it be exactly fair to charge a canister for some subnet not having updated its routing table quickly enough).

Regardless, even if we didn't charge individual canisters for the messages in streams, we could still have said messages count against the subnet's message memory usage.